rc/filetype/go.kak

```diff
@@ -73,6 +78,23 @@ evaluate-commands %sh{
 # Commands
 # ‾‾‾‾‾‾‾‾
 
+define-command go-alternative-file -docstring 'Jump to the alternate file (implementation ↔ test)' %{ evaluate-commands %sh{
+    case $kak_buffile in
+        *_test.go)
+            altfile=$(echo $kak_buffile | sed 's/_test\.go$/.go/')
+            test ! -f $altfile && echo "fail 'implementation file not found'" && exit
+        ;;
+        *.go)
+            altfile=$(echo $kak_buffile | sed 's/\.go$/_test.go/')
+            test ! -f $altfile && echo "fail 'test file not found'" && exit
+        ;;
+        *)
+            echo "fail 'alternative file not found'" && exit
+        ;;
+    esac
+    echo "edit $altfile"
```

Comment:
quoting altfile would be safer, use printf to avoid portability issues with echo.

Reply:
What portability issues are there here other than if `$altfile` contains backslashes? And how should this be quoted? If I use single quotes, the filename could break out of that with another; if I use braces, the filename could break out of that with a closing brace. This is overall handled very inconsistently through the rest of the codebase, here's a quick grep:

- `echo`, no quoting, no newline:
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/filetype/ruby.kak#L141
- `printf`, no quoting, newline
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/filetype/lua.kak#L85
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/filetype/moon.kak#L77
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/tools/go/go-tools.kak#L157-L160
- `printf`, single quotes, no newline
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/tools/git.kak#L273-L280
- `printf`, single quotes, newline
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/tools/rust/racer.kak#L80
- `printf`, brace quotes, newline:
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/filetype/c-family.kak#L420
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/filetype/c-family.kak#L431
- `awk`, `%|...|` quoting, implicit newline
https://github.com/mawww/kakoune/blob/cfbff4e132c8d0aed8bcaf677b25f5ab3a9c3221/rc/tools/ctags.kak#L40-L64

AFAICT, newlines are completely unneeded for kakoune to run the command, and could be dropped. If you pick a common style to use here, I can add a commit that makes these consistent.

Reply:
[Backslashes and flags are an issue with `echo`](https://www.etalabs.net/sh_tricks.html).

The path should be quoted for Kakoune, e.g. `printf 'edit "%s"' "$(printf %s "${altfile}" | sed 's,",&&,g')"`.

There's a lot of technical debt in the codebase, with unquoted variables, calls to `echo`, or lazy `printf %s\\n 'static string'`… I try to fix them when I see them in my own PRs.

Feel free to create a commit that addresses all these, but be aware that it's not gratifying work ;)

Reply:
I'm aware of that page, and flags aren't relevant here. Is it really only possible to escape things properly by running stuff through sed first? You don't have a function for that or such?

Reply:
I think the quoting can also be done in a follow up.

> The path should be quoted for Kakoune, e.g. printf 'edit "%s"' "$(printf %s "${altfile}" | sed 's,",&&,g')".

I believe this should use single quotes so `%val{foo}` expansion does not take place.

> There's a lot of technical debt in the codebase, with unquoted variables

Yeah, this on rarely matters in practise but is totally avoidable.

One more thing I'd do is use `edit -- '%s'` so it works for files that start with `-`, though I guess I'd be okay with ignoring those, too.

> You don't have a function for that or such?

Not yet, that's #3340

This one is used sometimes:
`kakquote() { printf "%s" "$*" | sed "s/'/''/g; 1s/^/'/; \$s/\$/'/"; }`
using it will make it easy to refactor, once we have some kind of prelude.

There is a version without `sed` calls in [prelude.kak](https://github.com/alexherbo2/prelude.kak/blob/master/rc/prelude.sh) but I don't think that's warranted here.

Reply:
@krobelus Thanks a lot for that function! Dropped that in and it works like a charm.

Reply:
(Don't merge yet, I'm adding commits to clean up the other places I found.)

Reply:
Why declare a function that is made to escape a *list* of arguments, in stead of the simple&stupid command I posted above?

Reply:
Because it would be better to have a common function that's used to escape things used throughout the entire codebase. It's easier to grep for instances of `kakquote` than to find every possible way someone's manually written out escaping logic.

Reply:
The function is only used in that one scope, in a single location. And it complicates the pattern passed to `sed` for no visible reason, hence my remark.

Reply:
"using it will make it easy to refactor, once we have some kind of prelude"

Reply:
Actually, probably better to merge this for now, I'm busier than expected. I'm going to make an issue regarding consistent shell script styling across the codebase, since that would make it easier to know what the correct way to implement things would be and avoid bikeshedding.

Reply:
Yeah, the shell script changes can be non-obvious and should be discussed in a follow-up.
My argument that it's easier to refactor `kakquote` is kind of weak, since we'll have to inspect every use of `sed` anyway.
So maybe the simple solution by lenormf is a better fit for the script in its current form.
Anyway, it really doesn't matter.

Reply:
¯\\\_(ツ)\_/¯